{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "checksum": "0b4e7a6b11c3878d09bc6eb4198835a2",
     "grade": false,
     "grade_id": "cell-4ec91fd127f50aa9",
     "locked": true,
     "schema_version": 1,
     "solution": false
    }
   },
   "source": [
    "# Week 1 Lab: next generation sequencing and mutation hunting\n",
    "\n",
    "## Part 2\n",
    "**Due: Monday 4/8/19 9am**\n",
    "\n",
    "**Partner: ** (optional. You must both turn in the assignment. You must choose a different partner each week.)\n",
    "\n",
    "Skills covered: More UNIX commands, python scripting, variant calling, error sources\n",
    "\n",
    "In this lab we'll dig into the sequence alignments we generated last time to identify potential mutations that occurred in the patient but not eiter of her parents."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "checksum": "8f7975619cda3afa0381a89b0105e7b8",
     "grade": false,
     "grade_id": "cell-ee3a082e514b9144",
     "locked": true,
     "schema_version": 1,
     "solution": false
    }
   },
   "source": [
    "## 1. Preparing for SNP calling: pileup files\n",
    "\n",
    "The goal now is to go through our data, and for each position in the reference genome, see how many reads have a mutation at the same position. The SAM file is inconvenient for this, because the reads aren’t matched with the reference in a simple way (though they do contain all the information needed). The solution is to make an intermediate file type called an mpileup, because it goes through each position and “piles up” the reads, tabulating the number of bases that match or don’t match the reference.\n",
    "\n",
    "mpileup requires a sorted, indexed bam file. We will focus only on a specific region of the genome to make this go faster. Run `samtools mpileup` to see the usage information. Then run the basic command below. Be sure to edit the path to the reference genome file.\n",
    "\n",
    "```\n",
    "samtools mpileup -r chr6:128405804-128605805 -f hg19.fa NA12878_child.sorted.bam > NA12878_child.mpileup\n",
    "```\n",
    "\n",
    "See what the pileup looks like with `head -n 100`. These files will be more convenient for SNP calling, since they \"pileup\" all the bases sequenced at each position in the reference genome. Read about pileup format here: http://samtools.sourceforge.net/pileup.shtml.\n",
    "\n",
    "```\n",
    "head -n 100 NA12878_child.mpileup | more\n",
    "```\n",
    "\n",
    "Now look at the usage of `samtools mpileup` to determine how to run the command on multiple BAM files at once (child, father, and mother, in that order so everyone's files are consistent). Output the results to `trio.mpileup`.\n",
    "\n",
    "To examine the output, we'll learn a new UNIX command, `grep`, which is a useful tool for searching files for specific string patterns (kind of like \"ctrl+F\" or \"find\" functions in other tools). Try the command below:\n",
    "\n",
    "```\n",
    "cat trio.mpileup | grep 128419307\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "checksum": "b9d1391a4e3e0a39dd177f01f56a4d6c",
     "grade": false,
     "grade_id": "mpileup",
     "locked": true,
     "schema_version": 1,
     "solution": false
    }
   },
   "source": [
    "**Question 1 (2 pts):** Look at the usage of `samtools mpileup` to determine how to run the command on multiple BAM files at once (child, father, and mother, in that order so everyone's files are consistent). Output the results to `trio.mpileup`. Paste your command below."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "checksum": "9e69793fb89daad1b9e327078654fe37",
     "grade": true,
     "grade_id": "correct-mpileup",
     "locked": false,
     "points": 2,
     "schema_version": 1,
     "solution": true
    }
   },
   "source": [
    "samtools mpileup -r chr6:128405804-128605805 -f ../../../datasets/cs185s-sp19-public/genomes/hg19.fa NA12878_child.sorted.bam NA12891_father.sorted.bam NA12892_mother.sorted.bam > trio.mpileup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "checksum": "2fdc0ab835121c6fd8f7ba61137efc84",
     "grade": false,
     "grade_id": "checkpos",
     "locked": true,
     "schema_version": 1,
     "solution": false
    }
   },
   "source": [
    "**Question 2 (2 pts)**: How many reads does each smaple (child, mother, father) have covering the position chr6:128419307? What does it look like the genotype of each sample is? (TT, AT, or AA)?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "checksum": "288e5ae1eb6971b14971f3171dd2942b",
     "grade": true,
     "grade_id": "correct-checkpos",
     "locked": false,
     "points": 2,
     "schema_version": 1,
     "solution": true
    }
   },
   "source": [
    "Child has 56 reads. <br>\n",
    "Father has 56 reads. <br>\n",
    "Mother has 50 reads. <br>\n",
    "\n",
    "For all three sample (child, father, mother) the genotype looks like it is AT. There are almost half matches (.,) and half mismatches (A,a). The original reference base pair is T."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "checksum": "9fe5c84982056b3333e8e6c51c0a6bed",
     "grade": false,
     "grade_id": "cell-6fc7f6432f923210",
     "locked": true,
     "schema_version": 1,
     "solution": false
    }
   },
   "source": [
    "## 2. Plotting the coverage distribution\n",
    "\n",
    "The mpileup format makes it straightforward to plot the coverage distribution for each sample. We can simply pull out the columns of the file for each sample correponding to coverage. \n",
    "\n",
    "**Question 3 (5 pts)** Complete the code below to plot the distribution of coverage and compute the mean of the distribution for each sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "checksum": "306cfa97bfab5587212d4785db0d2e5d",
     "grade": false,
     "grade_id": "covdist",
     "locked": false,
     "schema_version": 1,
     "solution": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Populating the interactive namespace from numpy and matplotlib\n",
      "Mean child coverage: 50\n",
      "Mean father coverage: 51\n",
      "Mean mother coverage: 54\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjgAAADQCAYAAAAK/RswAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAHhlJREFUeJzt3Xu4HVWZ5/Hvr8P9Gi7RjiEYhACDPCOX01yUsdPSQJBuAz1gQ9saGXjSj4KAtiOR6R6cBpww2oC0Qk+4DIFGIyIMaYimIxDAFpAEIhBizCFECGQgmAARGjDwzh+1dqdyss85+5yza1+qfp/n2c+pWntV1Sp4T+U9tVatUkRgZmZmVia/1+4GmJmZmTWbExwzMzMrHSc4ZmZmVjpOcMzMzKx0nOCYmZlZ6TjBMTMzs9JxgmNmZmal4wTHzMzMSscJjpmZmZXOFu1uQKvtvvvuMWHChHY3wwq0aNGilyNiTLvb0Qkc79XgmM843quh0XivXIIzYcIEFi5c2O5mWIEk/brdbegUjvdqcMxnHO/V0Gi8u4vKrEkkfVHSEklPSvqepG0k7SXpYUnLJX1f0lap7tZpvTd9PyG3n6+m8mWSjsuVT05lvZKmt/4Mzcy6hxMcsyaQNA44B+iJiAOBUcCpwKXA5RExEVgHnJE2OQNYFxH7AJenekg6IG33QWAycJWkUZJGAd8BjgcOAE5Ldc3MrA4nOGbNswWwraQtgO2A1cDHgFvT97OAE9PylLRO+v5oSUrlsyPirYh4BugFDkuf3ohYERFvA7NTXTMzq8MJjlkTRMTzwDeBZ8kSm1eBRcArEbEhVVsFjEvL44Dn0rYbUv3d8uV9tumvfDOSpklaKGnhmjVrRn5yZmZdyAmOWRNI2oXsjspewPuA7cm6k/qK2ib9fDfU8s0LI2ZGRE9E9IwZU/kHa8ysopzgmDXHHwPPRMSaiPgdcBvwYWB06rIC2AN4IS2vAsYDpO93Btbmy/ts01+5mZnVUbnHxKtmwvS7hlR/5YwTCmpJ6T0LHCFpO+DfgKOBhcC9wMlkY2amAnek+nPS+oPp+3siIiTNAb4r6TKyO0ETgZ+T3cGZKGkv4Hmygch/0aJzK62h/H74d8O6XdXi3QmOWRNExMOSbgUeBTYAjwEzgbuA2ZIuTmXXpU2uA26S1Et25+bUtJ8lkm4Bnkr7OSsi3gGQdDYwj+wJresjYkmrzs/MrNs4wTFrkoi4ELiwT/EKsieg+tZ9Eziln/1cAlxSp3wuMHfkLTUzKz8nOGZmZl1qqMMQqsSDjM3MzKx0nOCYmZlZ6TjBMTMzs9JxgmNmZmal4wTHzMzMSscJjpmZmZWOHxO3TVRtpkszMysn38ExMzOz0nGCY2ZmZqXjBMfMzMxKxwmOmZmZlY4THDMzMysdJzhmZmZWOk5wzMzMrHSc4JiZ2bBIGi3pVkm/lLRU0pGSdpU0X9Ly9HOXVFeSrpTUK+lxSYfk9jM11V8uaWqu/FBJT6RtrpSkdpyndScnOGZmNlzfAn4cEfsDHwKWAtOBuyNiInB3Wgc4HpiYPtOAqwEk7QpcCBwOHAZcWEuKUp1pue0mt+CcrCSc4JiZ2ZBJ2gn4KHAdQES8HRGvAFOAWanaLODEtDwFuDEyDwGjJY0FjgPmR8TaiFgHzAcmp+92iogHIyKAG3P7MhuUExwzMxuODwBrgP8j6TFJ10raHnhvRKwGSD/fk+qPA57Lbb8qlQ1UvqpO+SYkTZO0UNLCNWvWNOfMrBSc4JiZ2XBsARwCXB0RBwOvs7E7qp5642diGOWbFkTMjIieiOgZM2bM4K22ynCCY2Zmw7EKWBURD6f1W8kSnhdT9xLp50u5+uNz2+8BvDBI+R51ys0aUmiC4xH2ZmblFBH/D3hO0n6p6GjgKWAOULtOTwXuSMtzgM+ka/0RwKupC2secKykXdK/B8cC89J36yUdka7tn8nty2xQRd/B8Qh7M7Py+gJws6THgYOArwMzgGMkLQeOSesAc4EVQC9wDfB5gIhYC1wEPJI+f5fKAD4HXJu2eRr4UQvOyUpii6J2nBth/1nIRtgDb0uaAkxK1WYBC4DzyY2wBx5Kd3/GprrzawEvqTbCfgFphH0qr42w9y+AmVkLRMRioKfOV0fXqRvAWf3s53rg+jrlC4EDR9hMq6gi7+B0xAh7MzMzq54iE5yOGGEPfozQzMysaopMcDpmhL0fIzQzM6uWwhIcj7A3MzOzdilskHFSG2G/Fdno+dPJkqpbJJ0BPAuckurOBT5ONlr+jVSXiFgrqTbCHjYfYX8DsC3Z4GIPMDYzM7NiExyPsDczM7N28EzGZk3iiS3NzDqHExyz5vHElmZmHcIJjlkT5Ca2vA6yiS0j4hWyCSxnpWqzyCajhNzElhHxEFCb2PI40sSWEbEOqE1sOZY0sWXqzr0xty8zM+vDCY5Zc3TMxJae98nMzAmOWbN0zMSWnvfJzMwJjlmzdMzElmZm5gTHrCk8saWZWWcpeqI/syrxxJZmZh3CCY5Zk3hiSzOzzuEuKjMzMysd38Exs1KZMP2udjfBzDqA7+CYmZlZ6TjBMTMzs9IZNMGRdIqkHdPy30i6Lf9iQLOyccxblTjerawauYPztxGxXtJRZO/JmUV6MaBZSTnmrUoc71ZKjSQ476SfJ5BNQ38HsFVxTTJrO8e8VYnj3UqpkQTneUn/G/gkMFfS1g1uZ9atHPNWJY53K6VGgviTZNPHT46IV4Bdgf9aaKvM2ssxb1XieLdSGjTBiYg3yN5587qkPYEtgV8W3TCzdnHMW5U43q2sBp3oT9IXgAuBF4F3U3EA/7HAdpm1jWPeqsTxbmXVyEzG5wL7RcRvim6MWYdwzFuVON6tlBoZg/Mc8GrRDTHrII55qxLHu5VSI3dwVgALJN0FvFUrjIjLCmuVWXs55q1KHO9WSo0kOM+mz1Z4bgSrBse8VYnj3Upp0AQnIv4HQJrKOyLit4W3yqyNHPNWJY53K6tG3kV1oKTHgCeBJZIWSfpg8U0zaw/HvFWJ493KqpFBxjOBL0XE+yPi/cBfA9cU2yyztnLMW5U43q2UGklwto+Ie2srEbEA2L6wFpm1n2PeqsTxbqXU0FNUkv4WuCmt/yXwTHFNMms7x7xViePdSqmROzj/BRgD3AbcnpZPL7JRZm3mmLcqGVG8Sxol6TFJd6b1vSQ9LGm5pO9L2iqVb53We9P3E3L7+GoqXybpuFz55FTWK2l6U87WKqORp6jWAee0oC1mHcExb1XShHg/F1gK7JTWLwUuj4jZkv4ROAO4Ov1cFxH7SDo11ftzSQcApwIfBN4H/ETSvmlf3wGOAVYBj0iaExFPjaCtViH9JjiSroiI8yT9M9l7STYREZ8otGVmLeaYtyppRrxL2gM4AbgE+JIkAR8D/iJVmQV8jSzBmZKWAW4Fvp3qTwFmR8RbwDOSeoHDUr3eiFiRjjU71XWCYw0Z6A5OrT/2m61oiFkHcMxblTQj3q8AvgLsmNZ3A16JiA1pfRUwLi2PI3stBBGxQdKrqf444KHcPvPbPNen/PC+DZA0DZgGsOeee47gVKxs+h2DExGL0uJBEXFf/gMc1OgB3D9r3aJZMW/WDUYa75L+BHgptx8A1TvUIN8NtXzTgoiZEdETET1jxowZpNVWJY0MMp5ap+yzQzhGrX+2ptY/OxFYR9YvC7n+WeDyVI8+/bOTgatS0jSKrH/2eOAA4LRU12ykRhrzZt1kuPH+EeATklYCs8m6pq4ARkuq9Q7sAbyQllcB4wHS9zsDa/Plfbbpr9ysIf0mOJJOS32ze0mak/vcC/ymkZ3n+mevTeu1/tlbU5VZwIlpeUpaJ31/dN/+2Yh4Bqj1zx5G6p+NiLfJfsGmNHriZn01I+bNusVI4z0ivhoRe0TEBLI/Qu+JiE8B9wInp2pTgTvS8hw2JlMnp/qRyk9Nd/H3AiYCPwceASamu/5bpWPMacKpW0UMNAbnZ8BqYHfg73Pl64HHG9x/2/tnwX201rBmxLxZtygq3s8HZku6GHgMuC6VXwfclAYRryVLWIiIJZJuIRs8vAE4KyLeAZB0NjAPGAVcHxFLRtAuq5h+E5yI+DXwa+DI4ew43z8raVKtuN6hBvmuv/J6d58265+FrI+WbDpyenp66tYxG2nMm3WTZsZ7mv14QVpewcanoPJ13gRO6Wf7S8iexOpbPheYO9L2WTUN1EX10/RzvaTXcp/1kl5rYN/un7Wu0oSY96B66xrNiHezTjbQU1RHpZ87RsROuc+OEbFTf9vltnf/rHWVkcZ84kH11hWaFO9mHWvQp6gk7S1p67Q8SdI5kkaP4Jjnk00I1Us2xibfP7tbKv8SMB2y/lmg1j/7Y1L/bBrHU+ufXQrc4v5Za4bhxrwH1Vs3KuAab9YRGnlM/IfAO5L2IUtC9gK+O5SDRMSCiPiTtLwiIg6LiH0i4pQ0eyUR8WZa3yd9vyK3/SURsXdE7BcRP8qVz42IfdN3m/Xfmg3TcGO+Nqj+3bTe8KB6ID+ovu/g+XEDlJuN1Iiv8WadqJEE5910AT4JuCIivgiMLbZZZm015JhXh0x6ltoyTdJCSQvXrFkzQKvNAF/jraQaSXB+J+k0svExd6ayLYtrklnbDSfmO2ZQvWd2tSHyNd5KqZEE53SyxwgviYhn0kDffyq2WWZtNeSY96B662K+xlspDTTRHwAR8ZSkLwP7SjoQWBYRM4pvmll7NDnmPemZdTRf462sBk1w0iR9s4CVZOMAxkuaGhH3F9s0s/YYacx70jPrJr7GW1kNmuCQTeF9bEQsA5C0L/A94NAiG2bWRo55qxLHu5VSI2NwtqwFPkBE/AoPQLNyc8xblTjerZQauYOzUNJ1wE1p/VPAogHqm3U7x7xViePdSqmRBOdzwFnAOWT9s/cDVxXZKLM2c8xblTjerZQaeYrqLUnfBu4mm6F1WZoq3qyUHPNWJY53K6tGnqI6AfhH4Gmy7H4vSX+Vf2WCWZk45q2eCdPvarjuyhknFNiS5nK8W1k1+hTVH0VEL2QvZgPuAhz8VlaOeasSx7uVUiNPUb1UC/xkBfBSQe0x6wSOeasSx7uVUiN3cJZImgvcQvZyv1OARyT9GUBE3FZg+8zawTFvVeJ4t1JqJMHZBngR+MO0vgbYFfhTsl8GB7+VjWPeqsTxbqXUyFNUp7eiIWadwjFvVeJ4t3rKMKi+kTE4ZmZmZl3FCY6ZmZmVjhMcMzMzK51BExxJf5Nb3rrY5pi1n2PeqsTxbmXV7yBjSV8BHgBOBi5OxQ8Ch7SgXTaAoQz+ssY55q1KHO9WdgM9RbWMbD6ED0h6AFgK7CZpv4hY1pLWmbWWY96qxPFupTZQF9U64AKgF5gEXJnKp0v6WcHtMmsHx7xViePdSm2gOziTgQuBvYHLgF8Ar3vOBCsxx7xViePdSq3fOzgRcUFEHA2sBP6JLBkaI+mnkv65Re0zaxnHvFWJ493KrpFXNcyLiEfI3k3yuYg4StLuRTfMrI0c81YljncrpUZe1fCV3OpnU9nLRTXIrN0c81YljvfO4ydlm2NIE/1FxC+KaohZJ3LMW5U43q1MPJOxmZmZlY4THDMzMysdJzhmZjZkksZLulfSUklLJJ2byneVNF/S8vRzl1QuSVdK6pX0uKRDcvuamuovlzQ1V36opCfSNldKUuvP1LqVExwzMxuODcBfR8R/AI4AzpJ0ADAduDsiJgJ3p3WA44GJ6TMNuBqyhIhsPp7DgcOAC2tJUaozLbfd5Bacl5VEI4+JD4uk8cCNwO8D7wIzI+JbKZi/D0wgm3/hkxGxLmXm3wI+DrwBfDYiHk37mgrUXgh3cUTMSuWHAjcA2wJzgXMjIoo6J9vUUEb6r5xxQoEtMbNWi4jVwOq0vF7SUmAcMIVsZmSAWcAC4PxUfmO6Rj8kabSksanu/IhYCyBpPjBZ0gJgp4h4MJXfCJwI/KgV52fdr8g7OM7uzcwqQNIE4GDgYeC9KfmpJUHvSdXGAc/lNluVygYqX1WnvO+xp0laKGnhmjVrmnE6VhKFJTgRsbp2ByYi1pO9yK2W3c9K1WaRZeSQy+4j4iGglt0fR8ruI2IdUMvux5Ky+/QXwY25fZmZWQtI2gH4IXBeRLw2UNU6ZTGM8k0LImZGRE9E9IwZM6aRJltFtGQMTjuz+3R8Z/hWKA+4tCqStCVZcnNzRNyWil9Mf4CSfr6UylcB43Ob7wG8MEj5HnXKzRpSeILT7uwenOFbS7hL1iolJdjXAUsj4rLcV3OAWmI+FbgjV/6ZlNwfAbya/sidBxwraZcU68eSvT5iNbBe0hHpWJ/J7ctsUIUmOM7urSrcJWsV9BHg08DHJC1On48DM4BjJC0HjknrkD0IsgLoBa4BPg+QBhdfBDySPn9XG3AMfA64Nm3zNB5gbENQ5FNUg2X3M9g8uz9b0myyv15fjYjVkuYBX8/9FXss8NWIWCtpffpL4GGy7P4fijofs0YN1CUrqSVdsmR3ethzzz1HdjJm/YiIn1L/TjrA0XXqB3BWP/u6Hri+TvlC4MARNNMqrLAEh43Z/ROSFqeyC8gSm1sknQE8C5ySvptL9oh4L9lj4qdDlt1LqmX3sHl2fwPZY+I/wtm9tVnfLtkBhskU2iULzATo6enxtAlmVkmFJTjO7q1qBuqSTXdvGu2SndSnfAHukjUzGxLPZGzWBB5waWbWWYrsojKrEnfJmpl1ECc4Zk3gLlkzs87iBMfMOt5Q3ntmZgYeg2NmZmYl5ATHzMzMSscJjpmZmZWOExwzMzMrHSc4ZmZmVjpOcMzMzKx0nOCYmZlZ6TjBMTMzs9JxgmNmZmal4wTHzMzMSscJjpmZmZWOExwzMzMrHSc4ZmZmVjpOcMzMzKx0nOCYmZlZ6TjBMTMzs9LZot0NMDMrmwnT72q47soZJxTYErPq8h0cMzMzKx0nOGZmZlY6TnDMzMysdJzgmJmZWel4kLGZmZkN21AG1UPrBtY7wekQQw0QMzMz65+7qMzMzKx0fAfHzMysYL5L33pOcKwlPPGZ9eULvpkVyV1UZmZmVjpOcMzMzKx0nOCYmZlZ6XR9giNpsqRlknolTW93e8yK5Hi3qnHM23B19SBjSaOA7wDHAKuARyTNiYin2tsys+brhnj3wGFrpk6Pecd7Z+vqBAc4DOiNiBUAkmYDU4DCgt8BbW3UtHh3HHcOP2E4oKbEvOO9s7Qq5rs9wRkHPJdbXwUc3reSpGnAtLT6W0nL+lTZHXi5kBY2purH36QNunTE+3r/iPfQmZoV79Bh/88r3gZosB0D/G5UNua7KN5brRTn3E/MNxTv3Z7gqE5ZbFYQMROY2e9OpIUR0dPMhg1F1Y/fKW3oAk2Jd+iM/95uQ+e1owMNGvPdEu+tVsVz7qvbBxmvAsbn1vcAXmhTW8yK5ni3qnHM27B1e4LzCDBR0l6StgJOBea0uU1mRXG8W9U45m3YurqLKiI2SDobmAeMAq6PiCXD2NWAtzdboOrHh85oQ0drYrxDZ/z3dhs26pR2dJQSXePboYrnvAlFbNaFb2ZmZtbVur2LyszMzGwzTnDMzMysdCqf4LR6GnBJ4yXdK2mppCWSzk3lX5P0vKTF6fPxAtuwUtIT6TgLU9mukuZLWp5+7lLQsffLneNiSa9JOq+V519l7Zj2vhNiPteWtsV+Opbjv4Wq8pqHdsd1p6r0GJw0DfivyE0DDpxW5DTgksYCYyPiUUk7AouAE4FPAr+NiG8WdexcG1YCPRHxcq7sfwFrI2JGuhDsEhHnF9yOUcDzZBN3nU6Lzr+q2hHv6bhtj/lcW1bSAbGfjuv4L1C74r0dOimuO0nV7+D8+zTgEfE2UJsGvDARsToiHk3L64GlZLN1ttsUYFZankX2D1DRjgaejohft+BY1oZ4h46O+Zp2xD44/ovWlnjvIO2K645R9QSn3jTgLbvwSpoAHAw8nIrOlvS4pOsLvp0YwL9IWpSmOQd4b0SshuwfJOA9BR6/5lTge7n1Vp1/VbU13qGtMV/TKbEPjv+itT3eW6iT4rpjVD3BaWjq+0IOLO0A/BA4LyJeA64G9gYOAlYDf1/g4T8SEYcAxwNnSfpogceqK03a9QngB6moledfVW2Ld2h7zNe0PfbB8d8ibY33FuuIuO40VU9w2jINuKQtyS70N0fEbQAR8WJEvBMR7wLXkN1eLUREvJB+vgTcno71YhorURsz8VJRx0+OBx6NiBdTW1p2/hXWtmnv2x3zNR0S++D4b4XKvOahg+K6o1Q9wWn5NOCSBFwHLI2Iy3LlY3PVTgKeLOj426eBnkjaHjg2HWsOMDVVmwrcUcTxc04jd3u+VedfcW2Z9r7dMZ87XqfEPjj+W6ESr3nosLjuKJV+igogPY55BRunAb+k4OMdBTwAPAG8m4ovILvgHUR2C3Ul8Fe1/tMmH/8DZBk+ZK/q+G5EXCJpN+AWYE/gWeCUiFjb7OOnNmxH1jf+gYh4NZXdRAvOv+paHe/pmG2N+Vw72h77qR2O/xZpR7y3WqfEdSeqfIJjZmZm5VP1LiozMzMrISc4ZmZmVjpOcMzMzKx0nOCYmZlZ6TjBMTMzs9JxglMgSf9T0iRJJ7bqTbaSbpB0ckH7niTpzn6+u1bSAWl5paTd69T5mqQvF9E2a49ujnFJ+6e3Lz8mae8B6l2QW54gyXPUVEg3x/gI2zBa0udz6/1e/zuVE5xiHU72zp0/JJsHZECStii8RQWJiDPL+JZeG1Q3x/iJwB0RcXBEPD1AvQsG+G5IOuz8rTHdHOMjMRr4/KC1GtSO/y5OcAog6RuSHgf+AHgQOBO4WtJ/r1P3BkmXSboXuDTNSnm9pEfSX5ZTUr0Jkh6Q9Gj6fDiVS9K3JT0l6S5yL1STNCOVPy7pm0M8h30k/UTSL9Lxan/h7iDpVkm/lHRzmqUWSQsk9dTZz3+TtEzST4D9htIG61zdHuNpArjzgDNTu5D0f5W9rHCJ0gsLJc0Atk13em5Om4+SdE2q9y+Stk1195b047SPByTtX+/8h/Qf2tqm22M8bbtA0uWS7pe0VNIfSLpN0nJJF+fqfUnSk+lzXiqeAeydYv8bqay/6/+hku5LsT9PG18RsUDS1yXdB5w7lLY3RUT4U8CH7F0g/wBsCfzrAPVuAO4ERqX1rwN/mZZHA78Ctge2A7ZJ5ROBhWn5z4D5ZDN1vg94BTgZ2BVYxsbJHEcPsf0PAyel5W3S8ScBr5K90+X3yH7pj0p1FgA9aXklsDtwKNnstdsBOwG9wJfb/f/GH8d4qv+1fDwCu6af25JNdb9bWv9trs4EYANwUFq/JXcudwMT0/LhwD31zt+f7vmUIMYXAJem5XPJ3sU1Ftia7F1du+Wu09sDOwBLgINTrD+Z29ck6lz/03+bnwFjUr0/J5s1unb8q9r1/68st9I60cHAYmB/YLCumx9ExDtp+VjgE9o4VmUbsqm2XwC+Lekg4B1g3/T9R4Hvpe1fkHRPKn8NeBO4Nv1F0HDfqbL3moyLiNsBIuLNVA7w84hYldYXk/0S/LSfXf0n4PaIeCPVL917YCqua2O8H+dIOiktjyf7B+g3deo9ExGL0/IiYIKyN6V/GPhB+j2B7B+Rmvz5W/coQ4zXrrtPAEsivQJE0gqyOD+K7Dr9eiq/jezaXe96Xe/6/wpwIDA/xf4oIP+ake8Po81N4QSnyVLg3kCW5b5MlrErBcOREfFvdTZ7Pb8L4D9HxLI++/0a8CLwIbLs+c3c15u9byMiNkg6DDia7CVzZwMf67PPecB7yf6KOLNPG/rzVm75HQaPIb8LpGRKEuN9z2kS8Mep/W9IWkD2j1I9fX8Htk3tfSUiDupnm9f7KbcOVLIYr8Xru2wau++SXb8Hut73ty/YeP0XWeJ0ZD/btC32PQanySJicbrI/Qo4ALgHOC4iDurnl6KvecAXcn2bB6fynYHVEfEu8GmyLBngfuBUSaNSv+cfpe12AHaOiLlkYw02u/BGRK1dZ/Ypfw1YJenEtK+tlb0gcKjuB06StG26K/Snw9iHdZgyxHgdOwPrUnKzP3BE7rvfSdpyoI3T78wzkk5JbZOkDw1yTOtQJY3x/twPnChpO2VvIz+JbDD1emDHBrZfBoyRdGRq85aSPjjMtjSVE5wCSBpDdrF8F9g/hvZ00UVkfZqPK3sc9aJUfhUwVdJDZLc1a1nx7cBystuPVwP3pfIdgTuVDZK7D/jiEE/j02S37B8n61/9/SFuT0Q8SnZ7cjHwQxp4AsG6Q0liPO/HwBZpXxcBD+W+m5naenPdLTf6FHCGpF+QjWOYMoL2WJuVMMbrStfpG4Cfk429vDYiHouI3wD/mgYef2OA7d8mGy90aYr9xWTdtW3nt4mbmZlZ6fgOjpmZmZWOExwzMzMrHSc4ZmZmVjpOcMzMzKx0nOCYmZlZ6TjBMTMzs9JxgmNmZmal8/8BdKh51C7FksAAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 576x216 with 3 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "%pylab inline\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "# All output files will be stored in a week1 folder in your home directory\n",
    "BASEDIR = os.path.join(os.environ[\"HOME\"], \"week1\")\n",
    "\n",
    "# Note we used the pandas library to load the data file\n",
    "# covdata is a dataframe. use covdata[colname] to access a column\n",
    "# You can use numpy functions such as np.mean(covdata[colname])\n",
    "# to compute functions on columns\n",
    "# Use covdata.head() to view the top of the data frame.\n",
    "# See more at https://pandas.pydata.org/pandas-docs/stable/getting_started/tutorials.html\n",
    "\n",
    "def PlotCovDist(plot=False):\n",
    "    # Compute these values and return below\n",
    "    mean_child_cov = 0\n",
    "    mean_father_cov = 0\n",
    "    mean_mother_cov = 0\n",
    "    \n",
    "    # Load coverage ino\n",
    "    #covdata is a df with child, father, mother as column headers and each row is the number of reads at the bp\n",
    "    covdata = pd.read_csv(os.path.join(BASEDIR, \"trio.mpileup\"), sep=\"\\t\",\n",
    "                     names = [\"cov.child\", \"cov.father\", \"cov.mother\"],\n",
    "                     usecols = [3, 6, 9])\n",
    "\n",
    "    \n",
    "    # Plot coverage distributions\n",
    "    if plot:\n",
    "        fig = plt.figure()\n",
    "        fig.set_size_inches((8, 3))\n",
    "        ax = fig.add_subplot(131)\n",
    "        ax.hist(covdata[\"cov.child\"])\n",
    "        ax.set_xlabel(\"# reads - child\")\n",
    "        ax.set_ylabel(\"# positions\")\n",
    "        ax = fig.add_subplot(132)\n",
    "        ax.hist(covdata[\"cov.father\"])\n",
    "        ax.set_xlabel(\"# reads - father\")\n",
    "        ax.set_ylabel(\"# positions\")\n",
    "        ax = fig.add_subplot(133)\n",
    "        ax.hist(covdata[\"cov.mother\"])\n",
    "        ax.set_xlabel(\"# reads - mother\")\n",
    "        ax.set_ylabel(\"# positions\")    \n",
    "        fig.tight_layout();\n",
    "    \n",
    "    # Compute mean coverate of child, father, and mother\n",
    "    mean_child_cov = np.mean(covdata['cov.child'])\n",
    "    mean_father_cov = np.mean(covdata['cov.father'])\n",
    "    mean_mother_cov = np.mean(covdata['cov.mother'])\n",
    "\n",
    "    # Print out results\n",
    "    print(\"Mean child coverage: %d\"%mean_child_cov)\n",
    "    print(\"Mean father coverage: %d\"%mean_father_cov)\n",
    "    print(\"Mean mother coverage: %d\"%mean_mother_cov)\n",
    "\n",
    "    # Return results\n",
    "    return mean_child_cov, mean_father_cov, mean_mother_cov\n",
    "\n",
    "PlotCovDist(plot=True);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "checksum": "2104c509b138df50ef4ee1f5dc16e99b",
     "grade": true,
     "grade_id": "correct-covdist",
     "locked": true,
     "points": 5,
     "schema_version": 1,
     "solution": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mean child coverage: 50\n",
      "Mean father coverage: 51\n",
      "Mean mother coverage: 54\n"
     ]
    }
   ],
   "source": [
    "\"\"\"Check computed average coverages\"\"\"\n",
    "mean_child_cov, mean_father_cov, mean_mother_cov = PlotCovDist(plot=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "checksum": "a61d773fea29b14ca278d48b0f0005d4",
     "grade": false,
     "grade_id": "cell-72317cb114a6a737",
     "locked": true,
     "schema_version": 1,
     "solution": false
    }
   },
   "source": [
    "## 3. Running a SNP caller (Varscan)\n",
    "\n",
    "Now we will run a SNP caller to identify variants in our three samples. We could parse the mpileup file ourselves to do this, but there are already dozens of existing SNP callers out there so for now we'll choose not to reinvent the wheel.\n",
    "\n",
    "Until now, all the programs you have been using (BWA, fastqc, samtools...) were pre-installed for you in our public folder. Here you’ll get to try your hand at installing one type of program file.\n",
    "\n",
    "Use the command below (`curl` - it's a linux/unix command to access websites) to download VarScan into your `week1` folder. Notice that we are redirecting the contents of the download into a .jar file.\n",
    "\n",
    "```\n",
    "curl -L https://sourceforge.net/projects/varscan/files/VarScan.v2.3.9.jar/download > VarScan.jar\n",
    "```\n",
    "\n",
    "Try to bring up the man page with:\n",
    "```\n",
    "java -jar VarScan.jar\n",
    "```\n",
    "\n",
    "This install method won’t be the same for all bioinformatics programs, and it might not even be the same for other java programs, as installing open source software often has dependencies (prerequisite software) and can be sensitive to where its installed and how it is executed. We'll try to touch on this throughout the course.\n",
    "\n",
    "\n",
    "On the VarScan man page, you should see several commands. We are interested in the `mpileup2snp` command. Go ahead and enter it with `-h` to bring up the manual.\n",
    "\n",
    "```\n",
    "java -jar VarScan.jar mpileup2snp -h\n",
    "```\n",
    "\n",
    "There are lots of ways we could filter our data. VarScan lets the user define their own cutoffs for including data in calling variants, other programs rely on complex statistical procedures to evaluate the likelihood of real mutations.\n",
    "\n",
    "The first option we are interested in setting today is the `--min-var-frequency` option. This sets the minimum % of non-reference bases at a position required to call it a variant in the sample. Set this too low, and we'll pick up mostly sequencing errors. Set this too high, and we'll miss real variants. We're also interested in `--min-freq-for-hom`, the minimum % of non-reference bases to call a variant homozygous.\n",
    "\n",
    "Choose reasonable thresholds that will allow us to pick up both heterozygous and homozygous variants. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "checksum": "ee81a547b6a78b9f20e935da724da046",
     "grade": false,
     "grade_id": "varscan",
     "locked": true,
     "schema_version": 1,
     "solution": false
    }
   },
   "source": [
    "**Question 4 (4 pts)**: Choose reasonable thresholds to modify and run the command below to perform SNP calling. Paste your full command in the next cell. If you are unsure about the values you are putting in, you might want to discuss with your partner or TA since the results will affect what you do in the next part.\n",
    "\n",
    "```\n",
    "java -jar VarScan.jar mpileup2snp \\\n",
    "  trio.mpileup \\\n",
    "  --min-var-frequency N \\\n",
    "  --min-freq-for-hom N \\\n",
    "  --variants \\\n",
    "  --p-value 0.01 \\\n",
    "  --output-vcf 1 > trio.vcf\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "checksum": "7c240d6cb23ce921d6573722fcf26abf",
     "grade": true,
     "grade_id": "correct-varscan",
     "locked": false,
     "points": 4,
     "schema_version": 1,
     "solution": true
    }
   },
   "source": [
    "java -jar VarScan.jar mpileup2snp \\\n",
    "  trio.mpileup \\\n",
    "  --min-var-frequency 40 \\\n",
    "  --min-freq-for-hom  75 \\\n",
    "  --variants \\\n",
    "  --p-value 0.01 \\\n",
    "  --output-vcf 1 > trio.vcf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "checksum": "5a8779b17c1f70ec54e21c0c89f160a3",
     "grade": false,
     "grade_id": "cell-34612010b7238f37",
     "locked": true,
     "schema_version": 1,
     "solution": false
    }
   },
   "source": [
    "## 4. Intro to VCF files\n",
    "\n",
    "VarScan outputs variants in VCF (variant call format), a standard genomics format. To read about all the gory details of VCF files, see the specs: https://samtools.github.io/hts-specs/VCFv4.2.pdf\n",
    "\n",
    "First take a peak and scroll through the file:\n",
    "\n",
    "```\n",
    "cat trio.vcf | more\n",
    "```\n",
    "\n",
    "Key things to notice:\n",
    "\n",
    "* The header lines (starting with \"#\") give general information about the file such as the tool use to generate it and a description of all the different data fields in the file.\n",
    "* Below the header, there is one line per variant. For each variant, you can find the chromosome, position, reference allele, alternate allele, and additional info. Following these fields are the variant call and statistics for each sample\n",
    "* \"INFO\" fields describe per-variant statistics. \"FORMAT\" fields describe per-sample statistics at each variant.\n",
    "\n",
    "See lecture slides for more VCF info."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "checksum": "3728048f1df14db2ca0e8bce129e322f",
     "grade": false,
     "grade_id": "vcfcheck",
     "locked": true,
     "schema_version": 1,
     "solution": false
    }
   },
   "source": [
    "**Question 5 (4 pts)**: Look at the position chr6:128593590 (it might be helpful to use grep for this. See example command below). What is the genotype for each of the three samples? What is the p-value for each?\n",
    "\n",
    "```\n",
    "grep 128593590 trio.vcf\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "checksum": "d388a9da39ea8da3db484911558f01da",
     "grade": true,
     "grade_id": "correct-vcfcheck",
     "locked": false,
     "points": 4,
     "schema_version": 1,
     "solution": true
    }
   },
   "source": [
    "cat trio.vcf | grep 128593590 trio.vcf\n",
    "\n",
    "chr6    128593590       .       T       G       .       PASS <br>\n",
    "GT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR <br>\n",
    "0/0:126:68:68:68:0:0%:1E0:37:0:34:34:0:0 <br>\n",
    "0/0:85:45:45:45:0:0%:1E0:37:0:19:26:0:0 <br>\n",
    "0/1:70:52:52:32:20:38.46%:9.8643E-8:36:42:17:15:12:8 <br>\n",
    "\n",
    "Sample 1 Genotype: 0/0 (homozygous reference), pval = 1E0 <br>\n",
    "Sample 2 Genotype: 0/0 (homozygous reference), pval = 1E0 <br>\n",
    "Sample 3 Genotype: 0/1 (heterozygous), pval = 9.8643E-8 <br>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "checksum": "a355bc63224f45c50803209fe507e11a",
     "grade": false,
     "grade_id": "denovo",
     "locked": true,
     "schema_version": 1,
     "solution": false
    }
   },
   "source": [
    "## 5. Mutation identification and annotation\n",
    "\n",
    "We'd now like to identify potential *de novo* mutations that occurred in the child genome but that were not passed down by either parent. \n",
    "\n",
    "**Question 6 (10 pts)** In the space below, write a small program that identifies candidate mutations in the child genome.\n",
    "\n",
    "You may do this using whatever means you want, but example template code to get you started using the PyVCF library is provided. This can also be done using bash commands. Your code must run from within the notebook and output a list of candidate mutation positions.\n",
    "\n",
    "Note you could also do this manually, since our VCF file is pretty small. But we will give only 1 point for solutions with no code provided.\n",
    "\n",
    "The code prints out one line per position genotyped to give you an idea of how PyVCF works. Modify the code so that it returns a list of potentials of de novo mutations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "checksum": "c97cacf5cccbf838632ebb3fc41d7f45",
     "grade": false,
     "grade_id": "cell-d3188e9d674b4351",
     "locked": false,
     "schema_version": 1,
     "solution": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr6:128414945:C/T;C/T;C/T\n",
      "chr6:128419307:T/A;T/A;T/A\n",
      "chr6:128421762:C/T;C/T;C/T\n",
      "chr6:128435303:C/T;C/T;C/T\n",
      "chr6:128442104:G/G;G/G;G/T\n",
      "chr6:128446509:C/T;C/T;C/T\n",
      "chr6:128450019:T/G;T/G;T/T\n",
      "chr6:128465668:C/T;C/C;C/T\n",
      "chr6:128469937:T/A;T/A;T/A\n",
      "chr6:128471913:A/C;A/C;A/C\n",
      "chr6:128484504:A/G;A/G;A/A\n",
      "chr6:128486139:A/C;A/C;A/C\n",
      "chr6:128488195:T/C;T/C;T/C\n",
      "chr6:128493991:G/T;G/T;G/T\n",
      "chr6:128494093:C/G;C/G;C/G\n",
      "chr6:128496528:T/T;T/T;T/C\n",
      "chr6:128496557:C/T;C/T;C/T\n",
      "chr6:128503550:T/C;T/C;T/T\n",
      "chr6:128505804:A/C;A/A;A/A\n",
      "chr6:128514053:G/A;G/A;G/A\n",
      "chr6:128515718:T/C;T/C;T/T\n",
      "chr6:128517169:T/T;T/T;T/C\n",
      "chr6:128532763:G/C;G/C;G/C\n",
      "chr6:128559329:C/T;C/T;C/T\n",
      "chr6:128564669:G/T;G/T;G/T\n",
      "chr6:128568999:C/T;C/T;C/C\n",
      "chr6:128572877:C/A;C/A;C/A\n",
      "chr6:128583972:G/A;G/G;G/A\n",
      "chr6:128583983:T/G;T/T;T/G\n",
      "chr6:128585618:A/C;A/C;A/C\n",
      "chr6:128591818:T/G;T/G;T/G\n",
      "chr6:128593590:T/T;T/T;T/G\n",
      "chr6:128599580:C/A;C/A;C/A\n",
      "chr6:128599902:C/A;C/A;C/A\n",
      "chr6:128602090:A/C;A/C;A/C\n",
      "chr6:128602577:T/C;T/C;T/C\n",
      "chr6:128602796:C/T;C/T;C/T\n",
      "chr6:128603361:C/T;C/T;C/T\n",
      "chr6:128604078:C/T;C/T;C/T\n",
      "chr6:128604430:G/A;G/A;G/A\n",
      "chr6:128604571:A/C;A/C;A/C\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[128505804]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This code simply prints out each position in the VCF file and the genotypes\n",
    "# of each sample. Modify to return a list of positions of potential de novo\n",
    "# mutations. That is, positions where the child has an allele not seen in \n",
    "# either parent.\n",
    "\n",
    "import vcf\n",
    "\n",
    "# This code is using the PyVCF library to parse VCFs\n",
    "# See more here: https://pyvcf.readthedocs.io/en/latest/INTRO.html\n",
    "# You shouldn't have to deal directly with any PyVCF functions\n",
    "# in your solution.\n",
    "\n",
    "def GetDenovoPositions(printpos=False):\n",
    "    reader = vcf.Reader(open(os.path.join(BASEDIR, \"trio.vcf\"), \"r\"))\n",
    "    positions = []\n",
    "    \n",
    "    for record in reader:\n",
    "        chrom  = record.CHROM # e.g. chr6\n",
    "        pos = record.POS # e.g. 128414945\n",
    "        child_gt = record.genotype(\"Sample1\").gt_bases # e.g. 'T/T'\n",
    "        father_gt = record.genotype(\"Sample2\").gt_bases # e.g. 'T/A'\n",
    "        mother_gt = record.genotype(\"Sample3\").gt_bases # e.g. 'T/T'\n",
    "\n",
    "        # If the position is denovo, add pos to positions\n",
    "        if (child_gt != mother_gt) and (child_gt != father_gt):\n",
    "            positions.append(pos)\n",
    "        \n",
    "        \n",
    "        if printpos:\n",
    "            print(\"%s:%s:%s;%s;%s\"%(chrom, pos, child_gt, father_gt, mother_gt))\n",
    "        \n",
    "    return positions\n",
    "\n",
    "GetDenovoPositions(printpos=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "checksum": "1d4d70f7dd114a1877ac0166b297acc7",
     "grade": true,
     "grade_id": "cell-e0b06a95370255a1",
     "locked": true,
     "points": 10,
     "schema_version": 1,
     "solution": false
    }
   },
   "outputs": [],
   "source": [
    "\"\"\" Check the list of positions returned\"\"\"\n",
    "positions = GetDenovoPositions(printpos=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "checksum": "8d367c14012713dea96ac3a6d9194dde",
     "grade": false,
     "grade_id": "browser",
     "locked": true,
     "schema_version": 1,
     "solution": false
    }
   },
   "source": [
    "Now let's figure out if the mutations identified are likely to have an impact on any particular genes. We'll use the UCSC Genome Browser (https://genome.ucsc.edu/) to help with this. Go to the genome browser home page. Then on the top menu bar, hover over \"Genomes\" and choose the hg19 reference genome.\n",
    "\n",
    "**Question 7 (5 pts)**: In the search bar, navigate to the position of each mutation. e.g. \"chr6:XXXX\" in the search bar. Does any mutation fall within a gene? If yes, determine whether it is predicted to change the amino acid structure (i.e. is this a synonymous variant? missense? nonsense?) of the protein and describe the change below. Make sure to determine which strand the gene is on! (forward vs. reverse)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "checksum": "369b90bf3ccf4976f1369bc3e9d249e6",
     "grade": true,
     "grade_id": "correct-browser",
     "locked": false,
     "points": 5,
     "schema_version": 1,
     "solution": true
    }
   },
   "source": [
    "Looking up the specific location on UCSD gene browser, position 128505804, falls within the gene PTPRK. The reference genome has the bp A while the variant in question is C. The current triplet that the position is a part of codes for the amino acid L (CUA -> Leucine). For the gene, it is on the reverse strand. If the variant is substituted instead the amino acid becomes R (CGA -> Arginine). This is a missense mutation."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "checksum": "8fa397fe18cab93a5e0510384e7b609d",
     "grade": false,
     "grade_id": "cell-4c2db6b1607cdbed",
     "locked": true,
     "schema_version": 1,
     "solution": false
    }
   },
   "source": [
    "# 6. Additional challenge questions\n",
    "\n",
    "Provide short answers for each question below. Please no more than 3-5 sentences for each."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "checksum": "e2bcc7a51ec5842bd23eccc880fe0e1f",
     "grade": false,
     "grade_id": "seqcov",
     "locked": true,
     "schema_version": 1,
     "solution": false
    }
   },
   "source": [
    "**Question 8 (5 pts)**: You sequenced the genome of a patient in your clinic with a suspected neurodevelopmental disorder and plotted the distribution of coverage across all bases of the genome. The mean coverage is around 40x, with most bases covered between 20-60 times. However you notice some bases are never covered (0 coverage) and other bases have extremely large coverages (1000x). Describe what could be happening at positions with extremely low or high coverage. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "checksum": "bd957131be48164301d48aab28b11fd3",
     "grade": true,
     "grade_id": "cell-468011b9b4311e6b",
     "locked": false,
     "points": 6,
     "schema_version": 1,
     "solution": true
    }
   },
   "source": [
    "There are many locations in the whole workflow where this error can come from. The first thing to look into would be the alignment of the fragments. There could be missaligned sequences that are congregating in similar spots causing some areas to have high reads as well as some having little to none. There could also be duplicate reads for parts of sequences as well as contamination/low quality in sequences causing the duplicates/misalignments. Another possiblity is the pcr amplification step."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true,
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "checksum": "64eb7ff00decace9a7e4415ca41cbf18",
     "grade": false,
     "grade_id": "tlen",
     "locked": true,
     "schema_version": 1,
     "solution": false
    }
   },
   "source": [
    "**Question 9 (5 pts)**: For the same individual's genome, you plot the template length distribution (similar to how we did in Tuesday's lab.) If you plot the distribution over the entire genome, you see a mean template length of around 300bp. However, when you zoom in on region of the genome you expect is involved in the patient's disorder, you see a bimodal plot, with a peak at the expected 300bp and another at around 350bp. What could explain this?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "checksum": "ec5f0f7c2017f75e125e800998dee4bd",
     "grade": true,
     "grade_id": "cell-5110f3005032b938",
     "locked": false,
     "points": 6,
     "schema_version": 1,
     "solution": true
    }
   },
   "source": [
    "This could be cause by the workflow of how the sequences/fragments are produced. Different methods of sequncing (Sanger) and the quality of those runs."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true,
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "checksum": "1f97245dfbaa27296371134fa6774529",
     "grade": false,
     "grade_id": "binom",
     "locked": true,
     "schema_version": 1,
     "solution": false
    }
   },
   "source": [
    "**Question 9 (5 pts)**: You are sequencing the genome of a tetraploid genome (4 copies of the genome, rather than 2 like diploid humans). At a position with 30x coverage where only one of the four copies of the genome contains a variant from the reference genome, how many reads do you expect on average to have the variant?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "checksum": "cf6c7db2a6615ad4018a94ffdf89b3ba",
     "grade": true,
     "grade_id": "correct-binom",
     "locked": false,
     "points": 6,
     "schema_version": 1,
     "solution": true
    }
   },
   "source": [
    "You would expect 7-8 of the reads to contain the variant. This is cound by 30(total reads)/4(tetraploid) = 7.5 or 25% of the reads."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "checksum": "c07f8abef933de2f003e6b2fe7895bc0",
     "grade": false,
     "grade_id": "cell-96565e7ea0a57709",
     "locked": true,
     "schema_version": 1,
     "solution": false
    }
   },
   "source": [
    "## 7. Submit the assignment!\n",
    "\n",
    "Under the \"Assignment\" tab in JupyterHub, for the Lab1 assignment you should see buttons:\n",
    "\n",
    "* Validate: lets you check if your notebooks are passing all tests (including hidden tests)\n",
    "* Submit: will submit your assignment!\n",
    "\n",
    "Be sure to submit before lecture Monday (9am 4/8/19)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
